wgpu-core/src/validation.rs

```diff
@@ -762,6 +764,67 @@ pub fn check_texture_format(
     }
 }
 
+/// Returns an error if the target contains a second blend source and it is not
+/// supported by the device or shader endpoint
+pub fn check_target_blend_source<'a, A: crate::hal_api::HalApi>(
+    device: &crate::device::resource::Device<A>,
+    shader: Option<&crate::pipeline::ShaderModule<A>>,
+    fragment_state: &Option<crate::pipeline::FragmentState<'a>>,
+    state: &wgt::ColorTargetState,
+    target_index: u32,
+) -> Result<(), crate::pipeline::CreateRenderPipelineError> {
+    if let Some(blend_mode) = state.blend {
+        let mut got_second_blend_source = false;
+        if device
+            .features
+            .contains(wgt::Features::DUAL_SOURCE_BLENDING)
+        {
+            if let Some(fragment_state) = fragment_state.as_ref() {
+                if let Some(interface) = shader.and_then(|shader| shader.interface.as_ref()) {
+                    if let Some(entry_point) = interface.entry_points.get(&(
+                        naga::ShaderStage::Fragment,
+                        String::from(fragment_state.stage.entry_point.as_ref()),
+                    )) {
+                        for output in entry_point.outputs.iter() {
+                            if let &Varying::Local {
+                                ref location,
+                                ref iv,
+                            } = output
+                            {
+                                if *location == 0 && iv.second_blend_source {
+                                    got_second_blend_source = true;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        let check_target =
+            |factor: wgt::BlendFactor| -> Result<(), crate::pipeline::CreateRenderPipelineError> {
+                if factor.is_second_blend_source() {
+                    device.require_features(wgt::Features::DUAL_SOURCE_BLENDING)?;
+                    if target_index != 0 {
+                        return Err(crate::pipeline::CreateRenderPipelineError
+                            ::AlternateBlendSourceOnUnsupportedTarget { factor, target: target_index });
+                    }
+                    if !got_second_blend_source {
+                        return Err(crate::pipeline::CreateRenderPipelineError
+                            ::NonExistingSecondBlendSourceReferenced { factor });
+                    }
+                }
+                Ok(())
+            };
+        check_target(blend_mode.color.src_factor)?;
+        check_target(blend_mode.color.dst_factor)?;
+        check_target(blend_mode.alpha.src_factor)?;
+        check_target(blend_mode.alpha.dst_factor)?;
```

Comment:
This fn is only called for all shader outputs but there might be more render targets bound.

We should check if the factors reference the 2nd blend source [here](https://github.com/gfx-rs/wgpu/blob/e36c560748ece319a57e0c4a8fef1d1681ba20c5/wgpu-core/src/device/resource.rs#L2661). We'll end up with a bool `pipeline_expects_dual_source_blending` which we can later check against naga's `FunctionInfo::dual_source_blending` (avoiding the logic above this block since we already checked those invariants in naga).

Reply:
thanks. This looks like it is a bit more involved. I'll see if i can look at this tomorrow or next week.

Reply:
Is this resolved?

Reply:
I did upload a version with better validation logic, but I guess teoxoy should check if he is happy about it.